Metadata tag

kalaspuffar opened the pull request:
Issue #25.

Description:
Enable the addition of a metadata tag in OBFL to have it available in the PEF output.

Changes:
* Added a segment called metadata
* Added metadata map in RowDataProperties and RowImpl
* Added parsing of metadata tag
* Added functionality to print the metadata tag at the beginning of the first row of the block.

kalaspuffar committed:
Added metadata tag.

kalaspuffar committed:
Don't use object type.

kalaspuffar commented:
Hi @bertfrees and @PaulRambags

The tests will not pass until we merge the change in the API. When that is done I will work on this issue to fix eventual issues and run all tests. This PR is here for reference for now. If you want to look at the implementation and perhaps suggest improvements or changes.

Best regards
Daniel

bertfrees commented:
@kalaspuffar Looking good!

You may find it a detail, but I think it is important to make sure that if a metadata token directly precedes or follows a non-whitespace character, that is also reflected in the output. For example take the following input (the `|` is the token):

    ... foo |bar ...

If this text is split across two line in between the "foo" and the "bar", the result should be:

    .............. foo
    |bar

and not

    .............. foo |
    bar

I think your current implementation might be doing it wrong. It should be a simple adjustment.

kalaspuffar commented:
Hi @bertfrees 

I don't think my implementation changes any logic. I keep the information about a metadata tag in the block and then add it to the beginning of that block. Don't see that it could remove or change the spacing. But if you can find a use-case where this happens, I will stand corrected.

Best regards
Daniel

bertfrees commented:
If I understand correctly you are adding it to the beginning of the row, not the block (which is good).

I'm not saying you are changing any text or spacing, it's just about the position of the metadata token, i.e. the row it is attached to. I think in my example you are attaching it to the upper row. But I'll make a test so we can be sure.

kalaspuffar committed:
Changed to use external-reference tag with another namespace for attributes.

kalaspuffar commented:
Changes to this PR includes a new tag name and also added functionality to send information about new namespaces through the framework. If we don't add the namespaces used in OBFL the new PEF file will not be a correct XML document.

kalaspuffar committed:
Test fixes.

kalaspuffar committed:
Small test fix.

bertfrees commented on integrationtest/org/daisy/dotify/formatter/test/resource-files/flow-in/flow-in-footer-with-marker-reference-expected.pef:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<pef version="2008-1" xmlns="http://www.daisy.org/ns/2008/pef">
+<pef version="2008-1" xmlns="http://www.daisy.org/ns/2008/pef" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
```
Where does this come from? I know it's also in the source OBFL, but can't figure out what change in the code triggered this change in the PEF. Not wrong, just not so nice because the namespace isn't used in the PEF.

bertfrees replied:
Oh I see you are injecting fake `MetaDataItem` in the OBFL parser. Hmm I wonder if there is a better way to do it...
Also you only parse the namespace on the root element. That is not a sound solution.

bertfrees replied:
The trick here I think is to collect all the used namespaces (in `ObflParserImpl#parseExternalReference`, you should be able to do that with `StartElement#getNamespaceContext`), then, for each URI select one prefix (there could be more than one in theory), then use this info in `PEFMediaWriter` to create the namespace nodes on the `pef` element, and then when adding the attributes on `row` look up the right prefix in these namespace bindings.

bertfrees replied:
You might still have to use the `getMetaData()` trick to pass the info from the parser to the writer. I think it's really a hack but it might be fine if documented properly. Another solution could be to collect the namespace bindings on the writer side, but that means you can only start writing to the output stream when all the rows have been provided, which means it would not really be "streaming" anymore strictly speaking, don't think that's really a problem though. Of course another solution is to not put the namespace nodes on `pef` but on `volume`, `section`, `page` or `row`, which might result in a lot of them, which is also not so nice but not really a problem either.

kalaspuffar replied:
I would rather not put the namespaces on each row because that would be a lot of unnecessary overhead. And the hack to get the metadata from one document to the other seemed like a good compromise. Then again, I don't know if I documented it well enough. 

We could create a very sophisticated solution for this extra feature, or we could keep it light.
My suggestion is to try to document that when you need extra fields in a different namespace, you put that namespace on the root of the document to get them propagated. 

That would work with my solution and keep it simple.

Using a real XML writer for the PEFWriter in the future could be an excellent addition and changing this behavior than could be an option. I could add more if you like, but I don't want to get too far of scope for this solution.

bertfrees replied:
OK fair enough, we can keep the getMetadata trick to make it easier on the writer side, but it's really a total hack in my opinion, for several reasons, so needs to be documented very well. Still, on the parser side there is really nothing stopping us from parsing the XML correctly. What I suggested above is very simple to do.

kalaspuffar replied:
The extra namespace should now be removed and I've implemented the meta fetch in the parseExternalReference instead.

bertfrees commented on src/org/daisy/dotify/formatter/impl/segment/ExternalReferenceSegment.java:
```diff
@@ -0,0 +1,39 @@
+
+/**
+ * This segment contains all attributes from the optional
+ * metadata tags that could be added to blocks in order to
```
The "metadata" here is probably a remainder of your first version.

kalaspuffar replied:
You are right.

bertfrees commented on src/org/daisy/dotify/formatter/impl/row/RowImpl.java:
```diff
@@ -374,6 +391,15 @@ public boolean allowsBreakAfter() {
     }
 
+    /**
+     * Get the current metadata information tagged for this row.
```
"metadata"

bertfrees commented on test/org/daisy/dotify/formatter/impl/writer/PEFMediaWriterTest.java:
```diff
@@ -66,4 +71,43 @@ public void write(int b) throws IOException {
+        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd");
+        Date date = new Date(System.currentTimeMillis());
+
+        String exp = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
+                "<pef version=\"2008-1\" xmlns=\"http://www.daisy.org/ns/2008/pef\">" +
+                "<head>" +
+                "<meta xmlns:dc=\"http://purl.org/dc/elements/1.1/\">" +
+                "<dc:format>application/x-pef+xml</dc:format>" +
+                "<dc:identifier>identifier?</dc:identifier>" +
+                "<dc:date>" + formatter.format(date) + "</dc:date>" +
+                "</meta>" +
+                "</head>" +
+                "<body>" +
+                "<row example:id=\"TestValue\">Testing</row>" +
+                "</body>" +
+                "</pef>";
```
This doesn't look like valid XML? I guess this is just a unit test so could be fine, but it does show clearly what the problem is with the current implementation of PEFMediaWriter.

kalaspuffar replied:
Sadly, yes. The current implementation of PEF media writer don't use a real XML writer.

bertfrees commented on test/org/daisy/dotify/formatter/impl/writer/PEFMediaWriterTest.java:
```diff
@@ -66,4 +71,43 @@ public void write(int b) throws IOException {
+        PEFMediaWriter p = new PEFMediaWriter(new Properties());
+
+        Map<QName, String> ref = new HashMap<>();
+        ref.put(new QName(null, "id", "example"), "TestValue");
```
Note that this is an attribute without namespace, which isn't allowed in a PEF. So the PEF writer should probably ignore these (or raise an error).

bertfrees commented on test/org/daisy/dotify/formatter/impl/obfl/resource-files/ws-test-input-17.xml:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<obfl xmlns="http://www.daisy.org/ns/2011/obfl" version="2011-1" xml:lang="und" hyphenate="false" xmlns:example="http://example.com">
```
It looks like you're currently not doing anything with this file.

kalaspuffar committed:
Update documentation.

bertfrees commented:
There were no integration tests yet so added some: https://github.com/brailleapps/dotify.formatter.impl/commit/a49f7d89e204a7c13c09816fcbd5d7f0d41116a6 (updated link)

kalaspuffar committed:
Updated doc and added namespace.

kalaspuffar committed:
Changed to fetch namespace information in when parsing external reference.

bertfrees commented on src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java:
```diff
@@ -1603,11 +1591,27 @@ private void parseExternalReference(
         fc.insertExternalReference(attrList);
     }
 
+    private boolean checkIfAlreadyContainsPrefix(List<MetaDataItem> meta, MetaDataItem newItem) {
```
Actually it's better to check for the namespace URIs, not the prefixes. It is OK to have two mappings from the same prefix to different URIs, or two mappings from different prefixes to the same URI. The PEF writer can cope with that. But when you don't create a metadata item for a namespace that is used in the OBFL (because used prefix already mapped to a different URI), the writer can also not declare it on the `pef` element (with a generated prefix). Then it will have to declare it on the `row` element.

kalaspuffar replied:
Hi @bertfrees

Sound like a reasonable change. Updated to use namespaceURI instead.

Best regards
Daniel

kalaspuffar committed:
Changed to test namespaceURI instead of localpart.

bertfrees commented on src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java:
```diff
@@ -1579,6 +1582,37 @@ private void parsePageNumber(
+        while (it.hasNext()) {
+            Attribute a = it.next();
+            MetaDataItem newItem = new MetaDataItem(
+                new QName(null, a.getName().getPrefix(), "xmlns"),
```
Note that the URI normally associated with "xmlns" is "http://www.w3.org/2000/xmlns/". Not so important what we use in this case, but just for your information.

kalaspuffar replied:
Not important but added it for completeness.

PaulRambags commented on src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java:
```diff
@@ -1579,6 +1582,37 @@ private void parsePageNumber(
+        while (it.hasNext()) {
+            Attribute a = it.next();
+            MetaDataItem newItem = new MetaDataItem(
+                new QName(null, a.getName().getPrefix(), "xmlns"),
```
What happens when`a.getName()` does not have a prefix?

bertfrees replied:
It is forbidden in OBFL so the parser should ignore the attribute when that happens.

PaulRambags commented on src/org/daisy/dotify/formatter/impl/row/RowDataProperties.java:
```diff
@@ -57,6 +59,7 @@
         private String underlineStyle = null;
 
         private ListItem listProps = null;
+        private Object externalReference;
```
Please add ` =  null`

PaulRambags commented on src/org/daisy/dotify/formatter/impl/row/RowDataProperties.java:
```diff
@@ -161,6 +165,11 @@ public Builder widows(int value) {
             return this;
         }
 
+        public Builder addExternalReference(Object externalReference) {
```
Please rename this method to `externalReference`

PaulRambags commented on src/org/daisy/dotify/formatter/impl/row/RowImpl.java:
```diff
@@ -201,6 +205,17 @@ public Builder addIdentifiers(List<String> refs) {
         }
 
+        /**
+         * Add a external reference object that will flow though the framework to the PEF file.
```
`a external` -> `an external`

PaulRambags commented on src/org/daisy/dotify/formatter/impl/row/SegmentProcessor.java:
```diff
@@ -467,6 +468,12 @@ public boolean hasSignificantContent() {
                 return Optional.empty();
+            case ExternalReference:
+                if (currentRow == null) {
+                    newCurrentRow(spc.getMargins().getLeftMargin(), spc.getMargins().getRightMargin());
```
To me it seems a bit tricky to create a new row here.
Instead of creating a new row, you could also introduce `groupExternalReferences` and treat these in a similar way as the `groupMarkers`, `groupAnchors`, and `groupIdentifiers` in this file.

kalaspuffar replied:
Did some changes and the new implementation follows the intent of your comment. Thanks.

PaulRambags commented on test/org/daisy/dotify/formatter/impl/writer/PEFMediaWriterTest.java:
```diff
@@ -66,4 +71,43 @@ public void write(int b) throws IOException {
+                "</meta>" +
+                "</head>" +
+                "<body>" +
+                "<row example:id=\"TestValue\">Testing</row>" +
```
Where is the declaration of namespace `example`?

kalaspuffar replied:
Missing because I don't send in the example into the metadata parameter. Incorrect XML but still testing the feature.

PaulRambags replied:
>Missing because I don't send in the example into the metadata parameter.

This is a new feature that you've implemented in this PR, isn't it? So can't you fix it?

PaulRambags commented on src/org/daisy/dotify/formatter/impl/writer/PEFMediaWriter.java:
```diff
@@ -72,7 +73,7 @@ public void open(OutputStream os) throws PagedMediaWriterException {
     }
 
-    private void open(OutputStream os, List<MetaDataItem> data) throws PagedMediaWriterException {
+    public void open(OutputStream os, List<MetaDataItem> data) throws PagedMediaWriterException {
```
Please revert this change.
If you made it public because you want to test this method, please test the method above this one or another public method that uses this method.

kalaspuffar replied:
Could not revert this change as this is the way we add the namespaces to the output. I did not change this function because I wanted to test it. I changed it because we use it in our implementation and the test verifies that it works.

bertfrees replied:
@kalaspuffar I guess I didn't notice this change before. I don't understand. How exactly are you using this class?

kalaspuffar replied:
My bad made function private and removed test.

kalaspuffar commented:
Hi @bertfrees and @PaulRambags 

Did the changes you requested as I understood them. Also, did some small updates to the Javadoc, so we don't get those warnings anymore.

Please review it again when you have the time.

Best regards
Daniel

kalaspuffar committed:
Fixes from review and fixing some javadoc issues.

kalaspuffar committed:
Adding namespace to test in order to show implementation detail.

kalaspuffar committed:
Removed test and made function private.

kalaspuffar committed:
Removed imports

PaulRambags commented on src/org/daisy/dotify/formatter/impl/row/SegmentProcessor.java:
```diff
@@ -116,6 +119,7 @@
         this.groupMarkers = new ArrayList<>(template.groupMarkers);
         this.groupIdentifiers = new ArrayList<>(template.groupIdentifiers);
         this.leaderManager = new LeaderManager(template.leaderManager);
+        this.externalReference = template.externalReference;
```
Why at most one?
I'd expect here `this.externalReferences = new ArrayList<>(template.externalReference);`

kalaspuffar replied:
An external reference object is an object that could contain multiple values. Later, when we want to implement Berts' more extensive use-case, we might need more than one. But in this case, where we add attributes to a row, there can only be one object that can contain whatever you like. In this implementation, it is a map, but you can use the API to implement something else.

PaulRambags replied:
Can you have more than one `external reference` object in a single row?

kalaspuffar replied:
No, you can only have one external reference object per block in my implementation.

PaulRambags replied:
>No, you can only have one external reference object per block in my implementation.

Please see my remark below at `case ExternalReference:`

bertfrees commented:
Gonna be away from computer for a couple of days. @kalaspuffar Have you seen the tests I made (https://github.com/mtmse/dotify.formatter.impl/pull/24#issuecomment-629175879)?

PaulRambags commented on test/org/daisy/dotify/formatter/impl/writer/PEFMediaWriterTest.java:
```diff
@@ -66,4 +71,43 @@ public void write(int b) throws IOException {
+                "</meta>" +
+                "</head>" +
+                "<body>" +
+                "<row example:id=\"TestValue\">Testing</row>" +
```
>Missing because I don't send in the example into the metadata parameter.

This is a new feature that you've implemented in this PR, isn't it? So can't you fix it?

kalaspuffar replied:
Moved the test to an integrationtest instead which solves this issue.

kalaspuffar committed:
Moved test to use an integrationtest.

kalaspuffar committed:
Remove imports.

kalaspuffar commented:
Hi @bertfrees 

Yes, I saw your test cases and case 1 I could implement into my current solution. I wanted to add them using either cherry-pick or PR, but as they are not a part of this repository, it seemed impossible, so I've copied them over and added them to the PR.

For now, I have added ignore statements to test 2 and 3 as they require an implementation that handles more than one external reference per block. But we could extend the solution to address that in the future. Not in the scope of this implementation, though.

Best regards
Daniel

kalaspuffar committed:
Added intergration tests from bert.

kalaspuffar committed:
Update mockito to handle jdk11.

bertfrees commented:
Hi Daniel, you can cherry-pick with these commands:

    git remote add snaekobbi git@github.com:snaekobbi/dotify.formatter.impl.git && \
    git fetch snaekobbi && \
    git cherry-pick a49f7d8

Yes it's fine to skip some of the tests for now, although I would say test 2 and 3, but especially 2, are really not hard to fix so fixing them would only show thoroughness and good intentions.

kalaspuffar committed:
Don't keep result.

PaulRambags commented on src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java:
```diff
@@ -116,6 +116,7 @@
     Map<String, Node> xslts = new HashMap<>();
     Map<String, Node> fileRefs = new HashMap<>();
     Map<String, List<RendererInfo>> renderers = new HashMap<>();
+    private HashMap<QName, String> externalReferenceObject = null;
```
why is this map called 'externalReferenceObject' and not 'externalReferenceObjects`?
and what does it map? - this is not clear from the declaration.  Please use a clear name or add a comment that indicates what is being mapped here.

Personally I don't like to initate a hashmap with `null`. It is much nicer to initate it with an empty map.
In this way you can also abstract from the fact that it is a HashMap. It can be any map.
`private Map<QName, String> chooseABetterName = new HashMap<>();`

kalaspuffar replied:
This is the externalReferenceObject that could be anything but in this specific implementation seems to be a map. In order to check when it hasn't been initialized we use null as we don't want any extra data sent if not used.

PaulRambags replied:
Maybe I'm just too critical. But I will try to explain in more detail what I don't like about this statement.

1. It is unclear what is being mapped here. It maps a QName to a String. What does the String represent? That is not clear from the name `externalReferenceObject` and therefore there should be a comment that describes what is being mapped, especially, what the target of the mapping is.
2. Usually when you declare a map, you abstract from the implementation. That is, normally you would have something like `private Map<...> nameOfMap = ...` instead of `private HashMap<...> nameOfMap = ...`. Please have a look at all the code in this repository and you will see that a map is always _declared_ as a Map, although it may be _implemented_ as a HashMap. A solution would be to have here: `private Map<QName, String> externalReferenceObject = (HashMap)null`. But that looks fishy to me, see the next point.
3. I think it is a glitch to initialize a map with `null`, in general. This should be avoided if possible.

bertfrees replied:
From the name "externalReferenceObject" it is pretty obvious to me that it is about the "external-reference" OBFL element, no? So what else can the map be than a set of attribute name/value pairs? I think the name is fine.

I agree with not using HashMap as the variable type, but I'm not sure that you will not find any other examples of this in this repository.

No opinion about initializing with null.

PaulRambags commented on src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java:
```diff
@@ -874,6 +875,8 @@ void parseBlock(
         FormatterCore fc,
         TextProperties tp
     ) throws XMLStreamException {
+        externalReferenceObject = null;
```
If you want to empty the map, please use `externalReferenceObject.clear();`

kalaspuffar replied:
I don't want an empty map. I want to remove the object.

kalaspuffar replied:
Just to clarify, this could be implemented as an object. This is not a map. Only my implementation uses a map other implementations might choose other data types.

PaulRambags commented on src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java:
```diff
@@ -1588,7 +1591,9 @@ private void parseExternalReference(
     ) throws XMLStreamException {
         Iterator<Attribute> it = event.asStartElement().getAttributes();
-        Map<QName, String> attrList = new HashMap<>();
+        if (externalReferenceObject == null) {
```
The checking on `null` should be removed everywhere

kalaspuffar replied:
Why?

PaulRambags replied:
Please see my remark on line 123, under 3.

PaulRambags commented on src/org/daisy/dotify/formatter/impl/row/SegmentProcessor.java:
```diff
@@ -467,6 +471,9 @@ public boolean hasSignificantContent() {
                 applyAfterLeader((IdentifierSegment) s);
                 return Optional.empty();
+            case ExternalReference:
+                externalReference = ((ExternalReferenceSegment) s).getExternalReference();
```
if `externalReference` has already a value (which can happen if you would have more than one external reference object in a block and which is perfectly legal by the OBFL definition) then an `UnsupportedOperationException` should be thrown, I think.

kalaspuffar replied:
Yes you are right that it might happen if you have two references on the same row but then they are still in the same block which means that they only have one composite object reference.

kalaspuffar commented:
It seems most of the comments from @PaulRambags are hinting that we should implement the full solution that Bert has proposed and that I've said could be another pull request where we support more than one reference per block or row. And in that case, each segment and their position needs to retained and kept so we can add the external-reference element or spans in the output PEF.

This implementation is on block level and only sets attributes on the first row of that block, which means the implementation can be simplified.

Other than that, the API and implementation are incredibly generic using an Object all through the framework, which means we can't make any assumptions that there will be a map, for instance.

PaulRambags commented:
>It seems most of the comments from @PaulRambags are hinting that we should implement the full solution that Bert has proposed

No, for me it is not important that the full solution is implemented. But if someone has more than one `externalReference` in one block, the system should not just process one of the two without any warning or error. There should be some message stating that something is wrong. I would prefer that an exception is thrown.

kalaspuffar commented:
Well, my belief is that it would never happen. As the parser merges them into one object and the last one is used. This means that we always will get all data. If we add your solution the only thing we will introduce is crash states when no error has occurred.

I rather have something that works most of the time and don't report things that it feels is iffy if it doesn't change the outcome.

bertfrees commented:
Hi guys, been away for a few days.

As far as I'm concerned the "full-blown" solution with character-level accuracy is not in scope for this PR. That will lead us too far and it's only a theoretical use case anyway. The tests that I added are also not related to that use case by the way, they don't expect any annotations in the PEF on character level.

However, whereas character-level accuracy would require significantly more work, row-level accuracy is just a small step and seems like a good compromise.

There is also a difference between collapsing all values of a certain attribute (with a certain name) within a row or block into one, or ignoring all but one attribute, even those with a different name. It will work for you but it just seems like a big shortcut.

Same with the white space issue (see one of my tests), it all works great when there is only a external-reference element at the beginning of a block, but as soon as you deviate from that specific usage, things start to break. Like Paul, I'm also in favor of disallowing input that does not match the expected usage, by throwing an error in the OBFL parser. I would be absolutely fine with that. Then at least it's clear that we did not attempt to support those usages.

Finally, I had two more tests in mind that I didn't have time to create yet. I will do that now.

bertfrees committed:
Handle edge cases.

bertfrees commented on integrationtest/org/daisy/dotify/formatter/test/resource-files/external-reference3-input.obfl:
```diff
@@ -0,0 +1,21 @@
+   </layout-master>
+   <sequence master="master">
+      <block>
+        ⠁⠁⠁⠁⠁⠁⠁⠁⠁⠁
```
To be sure that this tests the right thing, should be less than 10 characters. I had put 7 (see also corresponding PEF).

bertfrees commented:
Two more tests here: https://github.com/snaekobbi/dotify.formatter.impl/commit/e2a08c0

kalaspuffar committed:
Added failure state.

kalaspuffar commented:
Hi @bertfrees 

Great edge case tests. I've added them and also done some small code tweaks to handle these.

Moreover, I've added an exception if someone is trying to use the function in a not defined use-case. I don't want to add to many execution halting exceptions as we want to produce documents as long as it's possible, but this might be an exception where the user might like to be informed that they use the framework in a way that isn't implemented yet.

Best regards
Daniel

bertfrees commented:
Great @kalaspuffar, I think throwing the exception solves a lot of issues.

Now I will be completely satisfied if there would be a small note in `ObflParserImpl` and `PEFMediaWriter` saying that the `List<MetaDataItem>` objects are not only used to pass `meta` elements from OBFL to PEF, as intended, but also to pass along namespace binding information to make it easier for `PEFMediaWriter` to declare namespace prefixes on the root element.

bertfrees commented on src/org/daisy/dotify/formatter/impl/writer/PEFMediaWriter.java:
```diff
@@ -227,9 +243,30 @@ public void newRow(Row row) {
+                for (Map.Entry<QName, String> entry : metadata.entrySet()) {
+                    QName name = entry.getKey();
+
+                    String prefix = namespaces.get(name.getNamespaceURI());
+                    if (prefix == null) {
+                        prefix = "ns" + namespaceCounter++;
+                        pst.print(" xmlns:" + prefix + "=\"" + name.getNamespaceURI() + "\"");
+                    }
+
+                    pst.print(" " + prefix + ":" + name.getLocalPart() + "=\"" + entry.getValue() + "\"");
```
Looks good. With a bit of reorganization wouldn't it have been possible to reuse the `getNamespaces()` method above though? Haven't looked at it closely.

kalaspuffar replied:
These namespaces for the global state is already removed when that function is executed. It gets namespaces for the meta lines in the header.

bertfrees replied:
I know.

kalaspuffar committed:
Changes for tests.

kalaspuffar committed:
added disclaimer.

kalaspuffar committed:
Do not decrypt.

kalaspuffar committed:
Reset testing.

kalaspuffar committed:
Added full output for tests

kalaspuffar committed:
remove unsupported variables.

kalaspuffar commented:
Hi @bertfrees and @PaulRambags 

I've now added the disclaimer to both parser and writer and also added full test output for failing tests in order to find another instance of unsupported variables in one of my test cases so now the tests pass again.

Please review again.

Best regards
Daniel

bertfrees commented:
Yes, thanks very much. Already approved.

PaulRambags commented on integrationtest/org/daisy/dotify/formatter/test/ExternalReferenceTest.java:
```diff
@@ -0,0 +1,111 @@
+
+    @Ignore("Not implemented yet, only supporting references on block level")
+    @Test
+    public void testExternalReference1() throws
```
Please use a descriptive name or add a comment that describes what you test
Please do this at all tests

PaulRambags commented on integrationtest/org/daisy/dotify/formatter/test/ExternalReferenceTest.java:
```diff
@@ -0,0 +1,111 @@
+        testPEF(
+                "resource-files/external-reference2-input.obfl",
+                "resource-files/external-reference2-expected.pef",
+                true
```
true -> false

PaulRambags replied:
@kalaspuffar have you resolved this?

PaulRambags commented on integrationtest/org/daisy/dotify/formatter/test/ExternalReferenceTest.java:
```diff
@@ -0,0 +1,111 @@
+        testPEF(
+                "resource-files/external-reference3-input.obfl",
+                "resource-files/external-reference3-expected.pef",
+                true
```
true -> false

PaulRambags commented on integrationtest/org/daisy/dotify/formatter/test/ExternalReferenceTest.java:
```diff
@@ -0,0 +1,111 @@
+        testPEF(
+                "resource-files/external-reference4-input.obfl",
+                "resource-files/external-reference4-expected.pef",
+                true
```
true -> false

PaulRambags commented on integrationtest/org/daisy/dotify/formatter/test/ExternalReferenceTest.java:
```diff
@@ -0,0 +1,111 @@
+        testPEF(
+                "resource-files/external-reference5-input.obfl",
+                "resource-files/external-reference5-expected.pef",
+                true
```
true -> false

PaulRambags commented:
@kalaspuffar I have reviewed the PR again. Apart from some minor things, could you have a look at my comments in `OBFLParserImpl` on this statement:
`private HashMap<QName, String> externalReferenceObject = null;`
Thank you.

kalaspuffar committed:
Changed naming of tests.

kalaspuffar committed:
Back to 7 instances.

kalaspuffar committed:
Don't use implementations for declarations.

PaulRambags reviewed (approved):
I really think that a map should _never_ be initiated with _null_. You could solve this issue by regarding an empty map as an absent `externalReference` object. That would be nicer in my opinion.

That would mean that you do `fc.insertExternalReference(externalReferenceObject);` only when `externalReferenceObject` is not empty.

But I don't want to block this PR any longer, so I approve reluctantly.

kalaspuffar commented:
Hi @PaulRambags and @bertfrees 

This pull request is now approved and I will run the regression tests.

Best regards
Daniel

PS. @PaulRambags I know that some newer programming paradigms wants everything to be initialized and available at runtime, and all null values are bad and hurtful. Sadly I'm from the old programming style where wasting memory, garage collection time and performance is bad. Just different views on the same problem and in this case it leads to the same result. DS.

kalaspuffar committed:
Forgot one debug boolean.
